kimera_vio_ros/src/interfaces/backend_interface.cpp

```diff
@@ -0,0 +1,227 @@
+  //   publishImuBias(output);
+  // }
+  if (pointcloud_pub_->get_subscription_count() > 0) {
+    publishTimeHorizonPointCloud(output);
+  }
+}
+
+
+void BackendInterface::publishState(
+  const VIO::BackendOutput::Ptr & output) const
+{
+  CHECK(output);
+  // Get latest estimates for odometry.
+  const VIO::Timestamp & ts = output->timestamp_;
+  const gtsam::Pose3 & pose = output->W_State_Blkf_.pose_;
+  const gtsam::Rot3 & rotation = pose.rotation();
+  const gtsam::Quaternion & quaternion = rotation.toQuaternion();
+  const gtsam::Vector3 & velocity = output->W_State_Blkf_.velocity_;
+  const gtsam::Matrix6 & pose_cov =
+    gtsam::sub(output->state_covariance_lkf_, 0, 6, 0, 6);
+  const gtsam::Matrix3 & vel_cov =
+    gtsam::sub(output->state_covariance_lkf_, 6, 9, 6, 9);
+
+  // First publish odometry estimate
+  Odometry odometry_msg;
+
+  // Create header.
+  odometry_msg.header.stamp = rclcpp::Time(ts);
+  odometry_msg.header.frame_id = world_frame_id_;
+  odometry_msg.child_frame_id = base_link_frame_id_;
+
+  // Position
+  odometry_msg.pose.pose.position.x = pose.x();
+  odometry_msg.pose.pose.position.y = pose.y();
+  odometry_msg.pose.pose.position.z = pose.z();
+
+  // Orientation
+  odometry_msg.pose.pose.orientation.w = quaternion.w();
+  odometry_msg.pose.pose.orientation.x = quaternion.x();
+  odometry_msg.pose.pose.orientation.y = quaternion.y();
+  odometry_msg.pose.pose.orientation.z = quaternion.z();
+
+  // Remap covariance from GTSAM convention
+  // to odometry convention and fill in covariance
+  static const std::vector<int> remapping{3, 4, 5, 0, 1, 2};
+
+  // Position covariance first, angular covariance after
+  DCHECK_EQ(pose_cov.rows(), remapping.size());
+  DCHECK_EQ(pose_cov.rows() * pose_cov.cols(),
+    odometry_msg.pose.covariance.size());
+  for (int i = 0; i < pose_cov.rows(); i++) {
+    for (int j = 0; j < pose_cov.cols(); j++) {
+      odometry_msg.pose
+      .covariance[remapping[i] * pose_cov.cols() + remapping[j]] =
+        pose_cov(i, j);
+    }
+  }
+
+  // Linear velocities, trivial values for angular
+  const gtsam::Matrix3 & inversed_rotation = rotation.transpose();
+  const VIO::Vector3 velocity_body = inversed_rotation * velocity;
+  odometry_msg.twist.twist.linear.x = velocity_body(0);
+  odometry_msg.twist.twist.linear.y = velocity_body(1);
+  odometry_msg.twist.twist.linear.z = velocity_body(2);
+
+  // Velocity covariance: first linear
+  // and then angular (trivial values for angular)
+  const gtsam::Matrix3 vel_cov_body =
+    inversed_rotation.matrix() * vel_cov * rotation.matrix();
+  DCHECK_EQ(vel_cov_body.rows(), 3);
+  DCHECK_EQ(vel_cov_body.cols(), 3);
+  DCHECK_EQ(odometry_msg.twist.covariance.size(), 36);
+  for (int i = 0; i < vel_cov_body.rows(); i++) {
+    for (int j = 0; j < vel_cov_body.cols(); j++) {
+      odometry_msg.twist
+      .covariance[i * static_cast<int>(
+          sqrt(odometry_msg.twist.covariance.size())) +
+        j] = vel_cov_body(i, j);
```

Comment:
@ToniRV , In relation to https://github.com/MIT-SPARK/Kimera-VIO/issues/41 , I think if there was a way to convert twist.linear and twist.covariance of the odometry_msg (or velocity_body) into a different reference frame, the then the wrapper could do the abstraction of allowing the imu frame and base link frames to have a non-identity displacement transform. This would also be handy if the transform between the IMU and base-link was non-static, like on sterio-imu pan tilt unit. But it would be nice to keep such reference frame transform functaions down in the kimera-vio library for all wrappers to share.